fairscale/experimental/optim/dynamic_loss_scaler.py

```diff
@@ -0,0 +1,203 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+To prevent underflow or overflow of gradients, Dynamic Loss Scaler is used to
+dynamically scale up and down gradients by scaling the loss.
+"""
+
+
+from collections import defaultdict
+from enum import Enum
+
+
+import torch
+
+
+class OptState(Enum):
+    READY = 0
+    UNSCALED = 1
+    STEPPED = 2
+
+
+def _refresh_per_optimizer_state():
+    return OptState.READY
+
+
+class DynamicLossScaler(object):
+    """An instance ``scaler`` helps perform the steps of gradient scaling
+    conveniently.
+
+    * ``scaler.scale(loss)`` multiplies a given loss by ``scaler``'s current scale factor.
+    * ``scaler.step(optimizer)`` safely unscales gradients and calls ``optimizer.step()``.
+    * ``scaler.update()`` updates ``scaler``'s scale factor.
+    """
+
+    def __init__(
+        self,
+        init_scale=2.0 ** 15,
+        scale_factor=2.0,
+        scale_window=2000,
+        tolerance=0.0,
+        threshold=None,
+        min_loss_scale=1e-4,
+    ):
+        self.loss_scale = init_scale
+        self.scale_factor = scale_factor
+        self.scale_window = scale_window
+        self.tolerance = tolerance
+        self.threshold = threshold
+        self._iter = 0
+        self._last_overflow_iter = -1
+        self._last_rescale_iter = -1
+        self._overflows_since_rescale = 0
+        self.min_loss_scale = min_loss_scale
+        self._per_optimizer_states = defaultdict(_refresh_per_optimizer_state)
+        self._scale = None
+
+    def scale(self, outputs):
```

Comment:
nit, but can we try to type everything @tmarkstrum ? I know it clumsifies a bit the python syntax, but in my experience it does help when using the codebase from a distance, and modern IDEs are making good use of this

Reply:
incidentally I think that it removes the need for some details in the documentation, since it's already there in the code, so two birds & one stone so to say